Add support for custom filters in Scaffolder Page

jescalada opened the pull request:
## Hey, I just made a Pull Request!
Related issue: #27313 

Please let me know which tests I should fix up if necessary! Also, I'm not sure if the changeset is valid or if I should re-generate it.

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->
### Changelog
- Added support for overriding default filters in the Scaffolder page
- Added custom tags filter example to documentation

The example I added allows splitting the Tags filter into multiple separate filters based on user-defined categories. It makes it easier to filter through lengthy template collections.

![image](https://github.com/user-attachments/assets/2554d845-27dc-4a2a-802f-a9fa05b42345)

Overriding the filters also lets users filter on custom template attributes, as described in the [Catalog Customization docs](https://github.com/backstage/backstage/blob/master/docs/features/software-catalog/catalog-customization.md/#customize-filters).

When the filters are not overridden (via `overrideFilters` prop), the previously availble default filters will be shown instead.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Thank you!

awanlin reviewed (changes requested):
Thanks for the contribution @jescalada, I have a fairly fundamental comment that would should address.

jescalada commented on docs/features/software-templates/writing-custom-filters.md:
```diff
@@ -0,0 +1,145 @@
+
+## Writing a Custom Filter
+
+Writing a custom filter for the scaffolder is almost identical to [writing custom filters for the catalog](https://backstage.io/docs/features/software-catalog/catalog-customization/#customize-filters).
```
Hi @awanlin, I thought that perhaps users would expect filter customization to appear under the plugin docs for each plugin. Users working on the scaffolder specifically might have a hard time finding what they need under `software-catalog > catalog-customization`.

Should I extend the original `catalog-customization` to show users how to add custom filters for existing types (custom Tags picker) and keep the other one for new types (SecurityTier picker)? I can then remove the redundant examples in `writing-custom-filters.md` and leave only the link to `catalog-customization` page.

awanlin replied:
The only documentation here should be how to pass the filter to be used by the Scaffolder. This should simply point to the Catalog filter documentation. 

Ideally would like to avoid multiple examples for the same concept. We could probably just add a sentence to the Catalog filters explaining that this can be used to also provide custom filters for existing filters if they have alternative implementations they would like.

jescalada replied:
Thanks for the feedback @awanlin!

As per the recent changes requested by @benjdlambert, customizing Scaffolder filters requires passing them as children components to the `Route`, as explained in a comment below. Would this be enough reason to keep a separate documentation file and example? I'm not sure if those changes are good to go, but I'll be able to fix up the docs once I get some feedback. I'm aware we're going for consistency/simplicity here, just not sure how to reconcile both the requested changes! 😅

benjdlambert reviewed (changes requested):
Hey :wave: thanks for this!

Thinking that maybe adding a prop for this is perhaps not the way we want to go if we want to make this like the Catalog filters.

I think it might make sense to reuse the patterns for how the `ScaffolderFieldExtensions` and `ScaffolderLayout` JSX wrapping components work in the old extensibility system for now.

We've also got some work coming up to move more of the catalog page to the new frontend system, so it's something we're going to be revisiting.

jescalada commented:
Hi @benjdlambert, I just modified this PR so that it can support custom filters in the following form:

```tsx
<Route
    path="/create"
    element={
      <ScaffolderPage
        defaultPreviewTemplate={defaultPreviewTemplate}
        groups={[
          {
            title: 'Recommended',
            filter: entity =>
              entity?.metadata?.tags?.includes('recommended') ?? false,
          },
        ]}
      />
    }
>
  <ScaffolderTemplateFilter>
    <CustomEntityTagFilterComponent />
  </ScaffolderTemplateFilter>
  <ScaffolderFieldExtensions>
    <DelayingComponentFieldExtension />
  </ScaffolderFieldExtensions>
  <ScaffolderLayouts>
    <TwoColumnLayout />
    </ScaffolderLayouts>
</Route>
```

Unfortunately, I was not able to get the props through to the `TemplateListPage`, so I hard-coded some default values in the `CustomEntityTagFilterComponent` itself to get it to work.

Is this more in line with what you're looking for? I'm aware that eventually this would be refactored so that filters can be configured with a .yaml file instead, but this might be a step in the right direction.

This is the code for the `CustomEntityTagPicker.tsx` which I added (note that the props have default values as they aren't working correctly):

<details>  

<summary>Code Details</summary>  

```tsx
import React from 'react';
import { DefaultEntityFilters, EntityFilter, useEntityList } from '@backstage/plugin-catalog-react';
import { Entity } from '@backstage/catalog-model';
import Checkbox from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormGroup from '@material-ui/core/FormGroup';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import { scaffolderPlugin } from '@backstage/plugin-scaffolder';
import { createScaffolderFilter } from '@backstage/plugin-scaffolder-react';

export type EntityTagPickerProps = {
  label: string;
  allowedTags: string[];
};

class CustomEntityTagFilter implements EntityFilter {
  constructor(readonly values: string[]) { }

  filterEntity(entity: Entity): boolean {
    const entityTags = entity.metadata.tags ?? [];
    return this.values.every(tag => entityTags.includes(tag));
  }
}

export type CustomFilters = DefaultEntityFilters & {
  customTags: CustomEntityTagFilter;
}

const CustomEntityTagPicker = ({ label = "Languages", allowedTags = ["React", "Java", "Go"] }: EntityTagPickerProps) => {
  const {
    filters: { customTags },
    updateFilters,
  } = useEntityList<CustomFilters>();

  function onChange(value: string) {
    const newTags = customTags?.values.includes(value)
      ? customTags.values.filter(tag => tag !== value)
      : [...(customTags?.values ?? []), value];
    updateFilters({
      customTags: newTags.length
        ? new CustomEntityTagFilter(newTags)
        : undefined,
    });
  }
  const lowercaseTags = allowedTags.map(tag => tag.toLocaleLowerCase('en-US'));

  return (
    <Box display="flex" flexDirection="column">
      <FormControl component="fieldset">
        <Typography variant="button">{label}</Typography>
        <FormGroup>
          {lowercaseTags.map(tier => (
            <FormControlLabel
              key={tier}
              control={
                <Checkbox
                  checked={customTags?.values.includes(tier)}
                  onChange={() => onChange(tier)}
                />
              }
              label={`${tier.charAt(0).toLocaleUpperCase('en-US')}${tier.slice(1)}`}
            />
          ))}
        </FormGroup>
      </FormControl >
    </Box >
  );
};

export const CustomEntityTagFilterComponent = scaffolderPlugin.provide(
  createScaffolderFilter({
    name: 'customTags',
    component: CustomEntityTagPicker,
  })
)
```

</details>

Let me know if this looks good so I can amend the documentation. Thank you!

jescalada commented on docs/features/software-templates/writing-custom-filters.md:
```diff
@@ -0,0 +1,145 @@
+        },
+      ]}
+      {/* highlight-add-start */}
+      overrideFilters={
```
I just modified this so that we can pass in a custom component through the body of the plugin. However, this requires the component to be registered using the `createScaffolderFilter` function (the same way the layout/extensions work). Furthermore, the `TemplateListPage` still uses an `additionalFilters` prop to get the filters from the scaffolder Router via the `useCustomFilters` hook.

Let me know if this makes sense or not! 😃

awanlin commented:
Sorry @jescalada, this has been on my todo list to come back to for a long time now, will do so later today in the context of the docs.

@benjdlambert, could you help review the code side of these changes?

awanlin reviewed (changes requested):
Hi @jescalada, took a look at this again, sorry I didn't get to it yesterday. My comments still hold and have not been addressed yet. When you have time, no rush, can you give it another look over please?

benjdlambert commented:
Sorry for the slow updates here! We've spoke about this in the SIG a little bit, and I think that the consensus is that we might want to pause this work until there's some movement on the Frontend System support for filters in the Scaffolder instead and go down that route.

Will probably bring this up again on Friday if you want to join and chat about it @jescalada? :praY:
